File: app/src/main/java/com/example/notes/CardSourceImpl.java
package com.example.notes;

import static com.example.notes.Note.LENGTH_BEGIN;

import android.content.res.Resources;
import android.content.res.TypedArray;

import java.util.ArrayList;

public class CardSourceImpl implements CardsSource {

    public static ArrayList<CardData> dataSource;
    protected static int[] pictures_global;
    private final Resources resources;

    public CardSourceImpl(Resources resources) {
        this.resources = resources;
        dataSource = new ArrayList<>(LENGTH_BEGIN);
        pictures_global = getImageArray();

    }

    private int[] getImageArray() {
        TypedArray pictures = resources.obtainTypedArray(R.array.pictures);
        int length = pictures.length();
        int[] answer = new int[length];
        for (int i = 0; i < length; i++) {
            answer[i] = pictures.getResourceId(i, 0);
        }
        return answer;
    }

    @Override
    public CardData getCardData(int position) {
        return dataSource.get(position);
    }

    @Override
    public int size() {
        return dataSource.size();
    }

    @Override
    public void deleteCardData(int position) {
        dataSource.remove(position);
    }

    @Override
    public void updateCardData(int position, CardData cardData) {
        dataSource.set(position, cardData);
    }

    @Override
    public void addCardData(CardData cardData) {
        dataSource.add(cardData);
    }

    @Override
    public void clearCardData() {
        dataSource.clear();
    }

    @Override
    public void setNewData(ArrayList<CardData> dataSource) {
        CardSourceImpl.dataSource = dataSource;
    }

    @Override
    public ArrayList<CardData> getCardData() {
        return dataSource;
    }

}

File: app/src/main/java/com/example/notes/CardSourceImpl.kt
package com.example.notes

import android.content.res.Resources
import com.example.notes.CardsSource
import com.example.notes.CardSourceImpl
import com.example.notes.CardData
import android.content.res.TypedArray
import com.example.notes.R
import java.util.ArrayList

class CardSourceImpl(private val resources: Resources) : CardsSource {
    init {
        dataSource = ArrayList(Note.LENGTH_BEGIN)
        pictures_global = imageArray
    }

    private val imageArray: IntArray
        private get() {
            val pictures = resources.obtainTypedArray(R.array.pictures)
            val length = pictures.length()
            val answer = IntArray(length)
            for (i in 0 until length) {
                answer[i] = pictures.getResourceId(i, 0)
            }
            return answer
        }

    override fun getCardData(position: Int): CardData {
        return dataSource[position]
    }

    override fun size(): Int {
        return dataSource.size
    }

    override fun deleteCardData(position: Int) {
        dataSource.removeAt(position)
    }

    override fun updateCardData(position: Int, cardData: CardData) {
        dataSource[position] = cardData
    }

    override fun addCardData(cardData: CardData) {
        dataSource.add(cardData)
    }

    override fun clearCardData() {
        dataSource.clear()
    }

    override fun setNewData(dataSource: ArrayList<CardData>) {
        Companion.dataSource = dataSource
    }

    override fun getCardData(): ArrayList<CardData> {
        return dataSource
    }

    companion object {
        @JvmField
        var dataSource: ArrayList<CardData>
        @JvmField
        var pictures_global: IntArray
    }
}